sdk/core/System.ClientModel/src/KeyCredential.cs

```diff
@@ -1,21 +1,41 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+using System.Threading;
+
 namespace System.ClientModel;
 
-public class KeyCredential
+// Type is sealed given need to use atomicity in reading/writing credential
+// data.  We won't want a type to inherit, add credential data, and in so doing,
+// lose atomicity of get/update operations.
+public sealed class KeyCredential
 {
     private string _key;
 
-    public string Key => _key;
+    /// <summary>
+    /// Initializes a new instance of the <see cref="KeyCredential"/> class.
+    /// </summary>
+    /// <param name="key">Key to use to authenticate with the Azure service.</param>
+    /// <exception cref="System.ArgumentNullException">
+    /// Thrown when the <paramref name="key"/> is null.
+    /// </exception>
+    /// <exception cref="System.ArgumentException">
+    /// Thrown when the <paramref name="key"/> is empty.
+    /// </exception>
+#pragma warning disable CS8618 // Non-nullable field is uninitialized. Consider declaring as nullable.
+    public KeyCredential(string key) => Update(key);
+#pragma warning restore CS8618 // Non-nullable field is uninitialized. Consider declaring as nullable.
 
-    public KeyCredential(string key)
-    {
-        _key = key;
-    }
+    // Note: this is GetValue instead of GetKey to allow consistent naming
+    // across credential types.  For example, for NamedKeyCredential, we would have
+    // GetValues with two out params to enable atomicity in reading credential values.
+    public string GetValue() => Volatile.Read(ref _key);
```

Comment:
@tg-msft, this is what we discussed in the review, and NamedKeyCredential would use a parallel name `GetValues` with two out params instead of `GetValue` with a single return value.

Is this sufficient for the level of consistency we want for these types?

Reply:
I don't know if there was agreement on `GetValue` vs. `GetSecret` vs `GetData`, etc., but I think any of those names is fine for now.  Would it be worth changing to `UpdateValue` as well for symmetry?

Reply:
If it's all the same, I'd rather keep consistency with Azure.Core names where we already have them.